src/client/visualizations/line-chart/chart-line/chart-line.tsx

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2017-2018 Allegro.pl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as d3 from "d3";
+import { Dataset, Datum, NumberRange, PlywoodRange, TimeRange } from "plywood";
+import * as React from "react";
+import { Stage } from "../../../../common/models/stage/stage";
+import { concatTruthy, flatMap, Unary } from "../../../../common/utils/functional/functional";
+import "./chart-line.scss";
+
+export type Scale = d3.scale.Linear<number, number>;
+
+type Range = NumberRange | TimeRange;
+
+interface ChartLineProps {
+  // TODO: ContinuousScale does not have compatible call signatures (union of function getting number and Date)
+  xScale: any;
+  yScale: Scale;
+  getX: Unary<Datum, Range>;
+  getY: Unary<Datum, number>;
+  color?: string;
+  showArea: boolean;
+  dashed: boolean;
+  dataset: Dataset;
+  stage: Stage;
+}
+
+type DataPoint = [number, number];
+
+const stroke = (color: string, dashed: boolean): Pick<React.CSSProperties, "stroke" | "strokeDasharray"> => ({
+  stroke: color,
+  strokeDasharray: dashed ? "4 2" : undefined
+});
+
+function areDetached(a: PlywoodRange, b: PlywoodRange): boolean {
+  return a.end.valueOf() !== b.start.valueOf();
+}
+
+function prepareDataPoints(props: Pick<ChartLineProps, "dataset" | "getX" | "getY">): DataPoint[] {
```

Comment:
Extract, better name and tests. This function picks input range midpoint and adds points for missing data.

Reply:
Probably split into two functions?